### PR TITLE
Feature/bison datetime parser

### PR DESCRIPTION
**Pull request by Shylock-Hg**
Refactor https://github.com/vesoft-inc/nebula/pull/2770 by bison.

Close https://github.com/vesoft-inc/nebula/issues/2720
Close https://github.com/vesoft-inc/nebula/issues/2630

### CODE REVIEW

File: src/common/time/parser/datetime_parser.yy [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+
+%start datetime
+
+%%
+
+datetime
+  : date date_time_delimiter time opt_time_zone {
+    if (outputType != nebula::time::Type::kDateTime) {
+      throw DatetimeParser::syntax_error(@1, "Mismatched date time type.");
+    }
```
[the remark this answers is not in the document]
Comment (jievince) [reply src/common/time/parser/datetime_parser.yy]: I still think it's improper to check the expected type in the parser. The parser is intended to parse the given string to the corresponding data structure. For a date string, the parser parses it as a `Date`, for a datetime string, parse parses it as a `Datetime`. Parser really don't care what the argument type the functions called that they expect

[the remark this answers is not in the document]
Comment (Shylock-Hg) [reply src/common/time/parser/datetime_parser.yy]: We need Datetime for some computation.

[the remark this answers is not in the document]
Comment (jievince) [reply src/common/time/parser/datetime_parser.yy]: OK

File: src/common/time/parser/datetime_parser.yy [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+  }
+  ;
+
+date_time_delimiter
+  : KW_TIME_ID
+  | SPACE
+  ;
+
+date
+  : INTEGER NEGATIVE INTEGER NEGATIVE INTEGER {
```
[the remark this answers is not in the document]
Comment (jievince) [reply src/common/time/parser/datetime_parser.yy:118]: > minutes could also use regex to validate. you have used the parser why not let it do more thing?

Where is this check? I don't find it.

[the remark this answers is not in the document]
Comment (Shylock-Hg) [reply src/common/time/parser/datetime_parser.yy:118]: See validateDate function

File: src/common/time/parser/datetime_parser.yy [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+  }
+  | date {
+    if (outputType != nebula::time::Type::kDate) {
+      delete $1;
+      throw DatetimeParser::syntax_error(@1, "Mismatched date time type.");
+    }
+    $$ = new DateTime(*$1);
+    *output = $$;
+    delete $1;
+  }
```
Comment (czpmango) [inline src/common/time/parser/datetime_parser.yy:100]:
```
neo4j@neo4j> RETURN datetime("2021-11-13");
+------------------------+
| datetime("2021-11-13") |
+------------------------+
| 2021-11-13T00:00Z      |
+------------------------+
```

Comment (czpmango) [reply src/common/time/parser/datetime_parser.yy:100]: It seems that this behavior is inconsistent with `neo4j`.

File: tests/tck/features/mutate/InsertWithTimeType.feature [hunk 4]
```diff
@@ -107,6 +107,76 @@ Feature: Insert with time-dependent types
     Then the result should be, in any order:
       | edge_date.f_date | edge_date.f_time  | edge_date.f_datetime         |
       | '2017-03-04'     | '23:01:00.000000' | '2017-03-04T22:30:40.000000' |
+    # insert with timezone offset and microseconds
+    When try to execute query:
+      """
```
Comment (czpmango) [inline tests/tck/features/mutate/InsertWithTimeType.feature:112]: Nice pr! For full consideration of `cypher` compatibility, can you move some TCK tests from [https://github.com/opencypher/openCypher/tree/master/tck/features/expressions/temporal](https://github.com/opencypher/openCypher/tree/master/tck/features/expressions/temporal)?

Comment (Shylock-Hg) [reply tests/tck/features/mutate/InsertWithTimeType.feature:112]: Good catch! I will add it later.

File: src/common/time/parser/datetime_parser.yy (hunk 3 above)
Comment (Shylock-Hg) [reply src/common/time/parser/datetime_parser.yy:100]: It's conflict with date rule.

**Review by jievince — DISMISSED** [review]
LGTM

**Comment by jievince** [comment]
Please resolve the conflicts

**Comment by Shylock-Hg** [comment]
> Please resolve the conflicts

Done

**Review by jievince — APPROVED** [review]

File: tests/tck/features/mutate/InsertWithTimeType.feature (hunk 4 above)
Comment (Aiee) [reply tests/tck/features/mutate/InsertWithTimeType.feature:112]: You might open an issue for this so you won't forget.

**Review by Aiee — APPROVED** [review]

**Review by jievince — APPROVED** [review]

**Review by Aiee — APPROVED** [review]

**Comment by HarrisChu** [comment]
incompatible with before.
should info users that `timestamp` just means time in seconds

```
(root@nebula) [s1]> yield timestamp("2020-01-11T19:28:23")
+----------------------------------+
| timestamp("2020-01-11T19:28:23") |
+----------------------------------+
| 1578770903                       |
+----------------------------------+
Got 1 rows (time spent 3166/14662 us)

Tue, 11 Jan 2022 16:29:18 CST

(root@nebula) [s1]> yield timestamp("2020-01-11T19:28:23.123456")
+-----------------------------------------+
| timestamp("2020-01-11T19:28:23.123456") |
+-----------------------------------------+
| BAD_DATA                                |
+-----------------------------------------+
Got 1 rows (time spent 2434/14832 us)

Tue, 11 Jan 2022 16:29:28 CST
```

```
(root@nebula) [s1]> yield timestamp(1641890171222)
+--------------------------+
| timestamp(1641890171222) |
+--------------------------+
| BAD_DATA                 |
+--------------------------+
Got 1 rows (time spent 2637/13288 us)

Tue, 11 Jan 2022 16:36:51 CST

(root@nebula) [s1]> yield timestamp(1641890171)
+-----------------------+
| timestamp(1641890171) |
+-----------------------+
| 1641890171            |
+-----------------------+
Got 1 rows (time spent 2490/10609 us)

Tue, 11 Jan 2022 16:36:53 CST
```

**Comment by Shylock-Hg** [comment]
> incompatible with before. should info users that `timestamp` just means time in seconds
> 
> ```
> (root@nebula) [s1]> yield timestamp("2020-01-11T19:28:23")
> +----------------------------------+
> | timestamp("2020-01-11T19:28:23") |
> +----------------------------------+
> | 1578770903                       |
> +----------------------------------+
> Got 1 rows (time spent 3166/14662 us)
> 
> Tue, 11 Jan 2022 16:29:18 CST
> 
> (root@nebula) [s1]> yield timestamp("2020-01-11T19:28:23.123456")
> +-----------------------------------------+
> | timestamp("2020-01-11T19:28:23.123456") |
> +-----------------------------------------+
> | BAD_DATA                                |
> +-----------------------------------------+
> Got 1 rows (time spent 2434/14832 us)
> 
> Tue, 11 Jan 2022 16:29:28 CST
> ```
> 
> ```
> (root@nebula) [s1]> yield timestamp(1641890171222)
> +--------------------------+
> | timestamp(1641890171222) |
> +--------------------------+
> | BAD_DATA                 |
> +--------------------------+
> Got 1 rows (time spent 2637/13288 us)
> 
> Tue, 11 Jan 2022 16:36:51 CST
> 
> (root@nebula) [s1]> yield timestamp(1641890171)
> +-----------------------+
> | timestamp(1641890171) |
> +-----------------------+
> | 1641890171            |
> +-----------------------+
> Got 1 rows (time spent 2490/10609 us)
> 
> Tue, 11 Jan 2022 16:36:53 CST
> ```

This PR don't modify this.

**Pull request merged**